src/common/AndroidUtils.ts

```diff
@@ -536,20 +781,96 @@ export class AndroidUtils {
         portNumber: number
     ): Promise<void> {
         const quote = process.platform === WINDOWS_OS ? '"' : "'";
-        const bootCmd = `shell ${quote}while [[ -z $(getprop sys.boot_completed) ]]; do sleep 1; done;${quote}`;
-        const command = `${AndroidUtils.getAdbShellCommand()} -s emulator-${portNumber} wait-for-device ${bootCmd}`;
-        const timeout = PlatformConfig.androidConfig()
-            .deviceBootReadinessWaitTime;
+        const command = `wait-for-device shell ${quote}while [[ -z $(getprop sys.boot_completed) ]]; do sleep 1; done;${quote}`;
+        const timeout = PlatformConfig.androidConfig().deviceReadinessWaitTime;
 
         const waitUntilReadyPromise = CommonUtils.promiseWithTimeout(
             timeout,
-            CommonUtils.executeCommandAsync(command),
+            AndroidUtils.executeAdbCommand(command, portNumber),
             `Timeout waiting for emulator-${portNumber} to boot.`
         );
 
         return waitUntilReadyPromise.then(() => Promise.resolve());
     }
 
+    /**
+     * Attempts to wait for an Android virtual device to power off. It determines whether a device is powered off
+     * by continously running 'adb devices' command and see if this particular device is no longer in the list.
+     *
+     * @param portNumber The ADB port of the Android virtual device.
+     */
+    public static async waitUntilDeviceIsPoweredOff(
+        portNumber: number
+    ): Promise<void> {
+        const isWindows = process.platform === WINDOWS_OS;
+        const command = isWindows
+            ? `powershell -Command "while($(adb devices | findstr emulator-${portNumber})){ Start-Sleep -s 1 }"`
+            : `while [[ -n $(adb devices | grep emulator-${portNumber}) ]]; do sleep 1; done;`;
+        const timeout = PlatformConfig.androidConfig().deviceReadinessWaitTime;
+
+        const waitUntilReadyPromise = CommonUtils.promiseWithTimeout(
+            timeout,
+            CommonUtils.executeCommandAsync(command),
+            `Timeout waiting for emulator-${portNumber} to power off.`
+        );
+
+        return waitUntilReadyPromise.then(() => {
+            // It may often be the case that the device is removed from 'adb devices' list too quickly
+            // specially on Windows platform. So we also wait an exra 5 seconds.
+            return CommonUtils.delay(5000);
```

Comment:
This 5sec is measured totally unscientifically 😃  On my Mac & Windows laptops it seems to work just fine.

Reply:
Number of retries in `remountAsRootWritableSystem` and the delay here. I wonder if we should make it configurable through JSON or environment variable. 5 seconds is reasonable to me. Just thinking about the case if user wants to take control.

Reply:
For now we are the only ones that are consumers of this code and we can control those values through the parameters to `remountAsRootWritableSystem`. As for this 5sec delay I don't think we should expose it b/c it is really a bad thing that we're doing here. I still would like to investigate other options and eventually remove this 5sec sleep from our code.

Reply:
Me too. 😀 It's an area that would be great if we could react to whatever adversity exists on the other side (e.g. bringing it back up), and employing retry logic there too.

I'm thinking there's going to come a day where we want to retool this flow. Some of the centralization of code (e.g. the bullet-proofing in `executeAdbCommand()`) may actually be limiting us, and it might make more sense to decentralize / bubble up the `adb` failure fallout into individual methods, which can then tailor their responses to their specific scenarios.

Sure would be great if we had a reliable API for these actions. 😐